Compile/src/Producers/JERSmearer.cc

```diff
@@ -66,66 +67,49 @@ void JERSmearer::Produce(ZJetEvent const& event,
-                else if ((m_smearingMethod == JERSmearer::SmearingMethod::STOCHASTIC) || (!matchedGenJet)) {
-                    // match not found or stochastic method requested -> draw pT smearing factor from a normal distribution
-                    recoJets[iJet]->p4 *= 1 + std::normal_distribution<>(0, jetResolution)(m_randomNumberGenerator) * std::sqrt(std::max(jetResolutionScaleFactor * jetResolutionScaleFactor - 1, 0.0));
+    for (std::map<std::string, std::vector<std::shared_ptr<KJet>>>::const_iterator itlevel = product.m_correctedZJets.begin();
+         itlevel != product.m_correctedZJets.end();
+         ++itlevel) {
+
+        // shortcuts for convenience
+        auto& recoJets = product.m_correctedZJets[itlevel->first];
+        const auto& matchedGenJetIndices = product.m_matchedGenJets[itlevel->first];
+        // go through all reco jets
+        for (size_t iJet = 0; iJet < recoJets.size(); ++iJet) {
+            // retrieve the (relative) pT resolution and the resolution scale factor
+            double jetResolution = m_jetResolution->getResolution({
+                {JME::Binning::JetPt, recoJets[iJet]->p4.Pt()},
+                {JME::Binning::JetEta, recoJets[iJet]->p4.Eta()},
+                {JME::Binning::Rho, event.m_pileupDensity->rho}
+            });
+            double jetResolutionScaleFactor = m_jetResolutionScaleFactor->getScaleFactor({
+                {JME::Binning::JetEta, recoJets[iJet]->p4.Eta()}
+            }, Variation::NOMINAL);
+            // get and validate matched gen jet
+            const KLV* matchedGenJet = nullptr;
+            if (matchedGenJetIndices[iJet] >= 0) {
+                matchedGenJet = product.m_simpleGenJets[matchedGenJetIndices[iJet]];
+
+                // invalidate match if additional pT-criterion is not met
+                if (std::abs(recoJets[iJet]->p4.Pt() - matchedGenJet->p4.Pt()) > 3 * jetResolution * recoJets[iJet]->p4.Pt()) {
+                    matchedGenJet = nullptr;
                 }
             }
+
+            // compute and apply the pT smearing factor
+            double jecSmearFactor = 1.0;
+            if ((matchedGenJet) && (m_smearingMethod == JERSmearer::SmearingMethod::HYBRID)) {
+                // match found and hybrid method requested -> smear using scaling method
+                jecSmearFactor = 1 + (jetResolutionScaleFactor - 1) * (recoJets[iJet]->p4.Pt() - matchedGenJet->p4.Pt())/recoJets[iJet]->p4.Pt();
+            }
+            else if ((m_smearingMethod == JERSmearer::SmearingMethod::STOCHASTIC) || (!matchedGenJet)) {
+                // match not found or stochastic method requested -> draw pT smearing factor from a normal distribution
+                jecSmearFactor = 1 + std::normal_distribution<>(0, jetResolution)(m_randomNumberGenerator) * std::sqrt(std::max(jetResolutionScaleFactor * jetResolutionScaleFactor - 1, 0.0));
+            }
+
+            // apply factor (prevent negative values)
+            if (jecSmearFactor > 0)
```

Comment:
Is the behaviour to skip the jet if the smearing factor is smaller zero desired? Shouldn't they be treated in some way?

Reply:
It's just the most straightforward way. An alternative would be to pick as many random numbers as necessary until a positive scale factor is drawn. Not sure that's worth the hassle though, since negative values are rare enough as it is.

Reply:
In this case I would recommend to open an issue to not lose track about this. In principal it would be nice in this case to use an positive definite distributed pdf. I would advise against vetoing a random number, because this might lead to unwanted behaviour (convolutions of pseudo rnd-number pdfs).

Reply:
Feel free to open an issue if you like, although I think this is quite minor and can be quickly resolved in this PR.
The [JetResolution TWiki[(https://twiki.cern.ch/twiki/bin/view/CMS/JetResolution#Smearing_procedures) recommends truncating negative factors to zero, so that would be a third option (see comment below).

Reply:
This is a good solution. We can keep it like that until the official recommendations change.